File: hw7/141044027/141044027/src/pkg141044027/GTUMap.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package pkg141044027;
import java.security.InvalidParameterException;
import javafx.util.Pair;
import pkg141044027.GTUSet;

/**
 *
 * @author burki
 * @param <K>
 * @param <V>
 */
public class GTUMap<K,V> extends GTUSet<javafx.util.Pair<K,V>> {
    
    public GTUMap(){
     
    }

    
    public static void main(String[]args){
    }
    //bu fonksiyon referance dondurcek
    public V at(K k) throws InvalidParameterException {
        boolean flag=false;
        for (int i = 0; i < this.currentSize; i++) {
            if(this.getVal(i).getKey().equals(k)){
                flag=true;
            }
        }
        if(flag==true){
            for (int i = 0; i < this.currentSize; i++) {
                if(this.getVal(i).getKey().equals(k)){
                    return this.getVal(i).getValue();
                }
            }
        }
        else
            throw new InvalidParameterException("there is no key for at() function.");
    return this.value[0].getValue();
    }
    
    public final void erase(Object element) throws InvalidParameterException{
        boolean flag=false;
        for (int i = 0; i < this.currentSize; i++) {
            if(this.getVal(i).getKey().equals(element)){
                flag=true;
            }
        }
        if(flag==false) throw new InvalidParameterException("There is no element for erasing");
        
        int index ;
        for (index = 0; index < this.currentSize; ++index) {
            if(this.getVal(index).getKey().equals(element)){
                this.currentSize-=1;
                break;
            }
        }
        Pair<K,V>[] temp = (Pair[]) new Pair[this.currentSize];
        int tempindex=0;
        for (int i = 0; i < this.currentSize+1; ++i) {
            if(i==index){}
            else {
                temp[tempindex] = this.getVal(i);
                
                tempindex++;
            }
        }
        this.value=temp;
        
        
    }
    
    public GTUIterator<javafx.util.Pair<K,V>> find(Object element) throws InvalidParameterException{
        int index = 0; 
        for (int i = index ; i < this.currentSize; ++i) {
            if(this.getVal(i).getKey().equals(element)){
                index=i;
            }    
        }    
        if(index!=0)
            return new GTUIterator<javafx.util.Pair<K,V>>(this.value,index);
        else
            throw new InvalidParameterException("there is no element for find function ");
    }
    
    
}

File: hw6/141044027/141044027/GTUMap.cpp

#include "GTUMap.h"

namespace myspace{

    template<class K,class V>
    GTUMap<K,V>::GTUMap():GTUSet<pair<K,V>>(){
    }

    template <class K,class V>
    bool GTUMap<K,V>::empty() noexcept {
        return GTUSet<std::pair<K,V>>::empty();
    }

    template <class K,class V>
    int GTUMap<K,V>::size() const{
        return GTUSet<std::pair<K,V>>::size();
    }

    template <class K,class V>
    int GTUMap<K,V>::max_size() const{
        return GTUSet<std::pair<K,V>>::max_size();
    }

    template <class K,class V>
    bool GTUMap<K,V>::insert(pair<K, V> element) {
        return GTUSet<std::pair<K,V>>::insert(element);
    }

    template <class K,class V>
    void GTUMap<K,V>::erase(pair<K, V> element) {
        GTUSet<std::pair<K,V>>::erase(element);
    }

    template <class K,class V>
    void GTUMap<K,V>::clear() noexcept {
        this->value.reset(new pair<K,V>());
        this->Size=0;
    }

    template <class K,class V>
    GTUIterator<pair<K, V>> GTUMap<K,V>::find(pair<K, V> element) {
        return GTUSet<std::pair<K,V>>::find(element);
    }

    template <class K,class V>
    bool GTUMap<K,V>::count(pair<K, V> element) {
        return GTUSet<std::pair<K,V>>::count(element);
    }

    template <class K,class V>
    GTUIterator<pair<K, V>> GTUMap<K,V>::begin() const{
        return GTUSet<std::pair<K,V>>::begin();
    }

    template <class K,class V>
    GTUIterator<pair<K, V>> GTUMap<K,V>::end() const{
        return GTUSet<std::pair<K,V>>::end();
    }

    template <class K,class V>
    void GTUMap<K,V>::maxSizeKontroller() {
        GTUSet<std::pair<K,V>>::maxSizeKontroller();
    }

    template <class K,class V>
    V &GTUMap<K,V>::operator[](const K &k) {
        bool flag=false;
        for (int j = 0; j < this->size(); ++j) {
            if(this->value.get()[j].first==k){
                flag=true;
            }
        }
        if(flag==true){
            for (int i = 0; i < this->size(); ++i) {
                if(this->value.get()[i].first==k){
                    return this->value.get()[i].second;
                }
            }
        }
        else
            throw invalid_argument("there is no key for []operator \n");
    }


}